text/0000-enum-variant-types.md

```diff
@@ -0,0 +1,203 @@
+            let x = a; // ok, a: Sum::A
+        }
+        b @ Sum::B => {
+            // b: Sum::B
+        }
+        c @ Sum::C => {
+            // c: Sum::C
+        }
+    }
+}
+```
+
+In essence, a value of a variant is considered to be a value of the enclosing `enum` in every matter
+but pattern-matching.
+
+Explicitly casting to the `enum` type forgets the variant information.
+
+```rust
+let x: Sum = Sum::A(5); // x: Sum
+let Sum::A(y) = x; // error: refutable match
+```
+
+In all cases, the most specific type (i.e. the variant type if possible) is chosen by the type
+inference. However, this is entirely backwards-compatible, because `Variant` acts as `Adt` except in
+cases that were previously invalid (i.e. pattern-matching, where the extra typing information was
+previously unknown).
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- The loose distinction between the `enum` type and its variant types could be confusing to those unfamiliar with variant types. Error messages might specifically mention a variant type, which could
+be at odds with expectations. However, since they generally behave identically, this should not
+prove to be a significant problem.
+- As variant types need to include generic parameter information that is not necessarily included in
+their definitions, it will be necessary to include explicit type annotations more often than is
+typical. Although this is unfortunate, it is necessary to preserve all the desirable behaviour of
+variant types described here: namely complete backwards-compatibility precise type inference
+(e.g. allowing `x` in `let x = Sum::A;` to have type `Sum::A` without explicit type annotations).
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The advantages of this approach are:
+- It naturally allows variants to be treated as types, intuitively.
+- It doesn't require explicit type annotations to reap the benefits of variant types.
+- As variant types and enum types are represented identically, there are no coercion costs.
+- It doesn't require type fallback.
+- It doesn't require value-tracking or complex type system additions such as refinement types.
+
+One obvious alternative is to represent variant types differently to enum types and then coerce them
+when used as an enum. This could potentially reduce memory overhead for smaller variants
+(additionally no longer requiring the discriminant to be stored) and reduce the issue with providing
+irrelevant type parameters. However, it makes coercion more expensive and complex (as a variant
+could coerce to various enum types depending on the unspecified generic parameters).
+
+Variant types have [previously been proposed for Rust](https://github.com/rust-lang/rfcs/pull/1450).
+However, it used a more complex type inference procedure based on fallback and permitted fallible
+coercion to variant types. The method proposed here is implementationally simpler and more
+intuitive.
```

Comment:
I want to see a discussion of whether this presents any problems for GADTs (or adding dependent types).
In particular:

```rust
enum Foo<T> { Baz, Bar(..) where T: Add }
// If you pattern match on `Bar` you get out a proof that `T: Add`.

enum Foo { for<T: Add> Foo(T) }

enum Id<S, T> { for<U> Refl : Id<U, U> }
```

I don't think there's a problem, but I want to be sure and not be presented with such a problem later...

Reply:
I'll make a small note, but I think it should be clear from the fact that all type information is preserved that this is forwards compatible with such extensions.